Re-assign existing webhooks during user deletion

thejamescollins opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

1. When deleting an administrator user, and the **Attribute all content to** option is chosen on the WordPress delete user screen, any existing webhook(s) owned by the user being deleted are re-assigned to the other user.
This helps prevent `woocommerce_rest_cannot_view` webhook payload errors that occur after an administrator with existing webhooks is deleted.
2. When deleting an administrator user without the **Attribute all content to** option being chosen on the WordPress delete user screen, any existing webhook(s) owned by the user being deleted are re-assigned to `user_id` zero.
This avoids webhooks from being left in the database with an invalid (non existent) `user_id`.
An admin can edit and re-save each of these webhooks in order to get them to be assigned to a valid user. This is because re-saving a webhook that has user_id=0 causes the webhook ownership to be updated to the currently logged in user.
3. The "Delete users" WordPress page will display a text explaining that the users that are about to be deleted have webhooks assigned, if that's the case.
4. Adds the ability to search for webhooks matching a `user_id` using `WC_Webhook_Data_Store::search_webhooks()` by specifying a `user_id` argument.

Closes #37806.

### How to test the changes in this Pull Request:

For the original issue: follow the **Steps to reproduce** in #37806, however in the last step (step 12), the webhook payload will be a valid order payload rather than a `woocommerce_rest_cannot_view` error.

For the new behavior of the user deletion:

1. Create two administrator users (Users - Add New, select "Administrator" in "Role").
2. Login and create a new webhook (WooCommerce - Settings - Advanced - Webhooks) with each of the newly created users.
3. Go to the users page, check the box for the two newly created users, select "Delete" in the "Bulk actions" dropdown, then click "Apply".

You'll now see a message that warns  about existing webhooks right before the "Confirm Deletion" button:

![image](https://user-images.githubusercontent.com/937723/235643750-4309a3ac-fcd4-4302-8294-a7c5be09fa2d.png)

4. Choose the "Delete all content" radio button and click "Confirm Deletion".
5. Verify that the webhooks still exist and have user id 0 assigned. The easiest way is to directly query the database: `wp db query "select name,user_id from wp_wc_webhooks"`
6. Verify that editing and saving one of the webhooks with user id 0 will indeed set the user id of that webhook to the id of the user currently logged in.

If a user doesn't have any content assigned there will be no choice as to what to do with the user content. The webhooks warning will still appear but with a slightly different wording:

![image](https://user-images.githubusercontent.com/937723/235644734-c1bcac26-539d-47c9-8bb9-2ef4f97efb2d.png)

Testing this isn't straightforwards since a post is created automatically for an user the first time it logs in. You can try creating yet another administrator user, *not* logging in with that user, taking note of its user id (search for `user_id` in the query string of the user edit url in the users page) and manually modifying one of the entries in the `wp_wc_webhooks` table to have that value for `user_id`.

thejamescollins committed:
Add test

thejamescollins committed:
Add ability to search for webhooks matching a `user_id`

thejamescollins committed:
Whenever a user is deleted, re-assign their webhooks to the new user.

thejamescollins committed:
Add changelog

thejamescollins committed:
Add unit test for `WC_Webhook_Data_Store::search_webhooks()` changes

om4james commented:
> I wasn't sure what we should do in cases where an administrator user is deleted _without_ the **Attribute all content to** option being chosen.
> 
> In that case we still end up with `wp_wc_webhooks` database records with a `user_id` that is no longer valid, which then causes `woocommerce_rest_cannot_view` payloads whenever data is sent to those webhooks.

I'm very interested if anyone has any feedback on what we could or should do in that case.

Right now, sending a `WP_Error` style payload to the delivery URL is not ideal.

And once webhooks are in existence with a non-existent/invalid `user_id`, there is no easy way for a store owner or administrator to fix them without editing the database directly.

Thanks

thejamescollins committed:
Re-assign webhook ownership to user_id zero when user is deleted without re-assignment

This avoids webhooks from being left in the database with an invalid (non existent) `user_id`.

An admin can edit and re-save each of these webhooks in order to get them to be assigned to a valid user. This is because re-saving a webhook that has user_id=0 causes the webhook ownership to be updated to the currently logged in user.


Also add another test

thejamescollins committed:
Update changelog to reflect latest changes

om4james commented:
> I wasn't sure what we should do in cases where an administrator user is deleted _without_ the **Attribute all content to** option being chosen.
> 
> In that case we still end up with `wp_wc_webhooks` database records with a `user_id` that is no longer valid, which then causes `woocommerce_rest_cannot_view` payloads whenever data is sent to those webhooks.


In c56a6d4 I have elected to re-assign those webhooks to `user_id` zero.

This avoids webhooks from being left in the database with an invalid (non existent) `user_id`.

Any admin can edit and re-save each of these webhooks in order to get them to be assigned to a valid user. This is because re-saving a webhook that has `user_id`=`0` causes the webhook ownership to be updated to the currently logged in user.

Konamiman committed:
Move the wc_reassign_webhooks_to_new_user_id function to a new WebhookUtil class.

Konamiman reviewed (changes requested):
Hi @om4james, thanks for your contribution. I have a couple of changes to request.

First, [as per our documentation](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/includes#woocommerce-includes-files) (although we can agree that it isn't very easy to find) we try to not put any new code inside the `includes` directory, which we consider as legacy; all new code should go into `src` instead. Thus the `wc_reassign_webhooks_to_new_user_id` function that you created should go into a class in `src` instead of being a function.

I have done that change myself in my local development environment, but I don't have permission to push the changes into your fork of WooCommerce. If you want to [give WooCommerce maintainers push permissions for your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) I'll push the change.

Second, it would be great if the user deletion confirmation page had a message for users with webhooks assigned, explaining that these will either be assigned user id 0 or the user id of the selected "Attribute all content to" user. This message could be added by using the `delete_user_form` hook. I could do that myself too if you want, or I can leave the fun of doing it for yourself, whatever you prefer. 🙂

So, if you are willing to give me push permissions for your fork I'll be happy to push the first change (and work on the second change too if you want). Thank you!

om4james commented:
Hi there @Konamiman, thank you for your feedback!

> First, [as per our documentation](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/includes#woocommerce-includes-files) (although we can agree that it isn't very easy to find) we try to not put any new code inside the `includes` directory, which we consider as legacy; all new code should go into `src` instead. Thus the `wc_reassign_webhooks_to_new_user_id` function that you created should go into a class in `src` instead of being a function.

Thanks for pointing that out, I wasn't aware of that policy but it definitely makes sense.

> I have done that change myself in my local development environment, but I don't have permission to push the changes into your fork of WooCommerce. If you want to [give WooCommerce maintainers push permissions for your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) I'll push the change.

Unfortunately I don't seem to be able to give you permission to push to our fork. I think it's because it's an organization owned fork rather than a user owned fork.

I'm happy for you to contribute and make that change, but I'm not sure of the best way to do so given the organization owned fork limitation. Any suggestions welcome :)
 
> Second, it would be great if the user deletion confirmation page had a message for users with webhooks assigned, explaining that these will either be assigned user id 0 or the user id of the selected "Attribute all content to" user. This message could be added by using the `delete_user_form` hook. I could do that myself too if you want, or I can leave the fun of doing it for yourself, whatever you prefer. 🙂

I like that idea in principle, I think the challenge is that users generally have no idea at all that webhooks are "owned" by a user. Deleting a user and assigning product authorship to a new user is probably easily understood by a user, but warning users of webhook ownership transfer might confuse most users during the deletion process?
If we add a web hook-related message to the `delete_user_form`, then we should probably also do the same thing for products etc too?

At that point I wonder if it may also be out of the scope of this PR then too?

Thanks again for your input and time so far!

om4james commented:
Hi @Konamiman,

I just invited you to have write access to our https://github.com/OM4/woocommerce-develop fork.

I think this should mean that you can now push to [the branch that this PR is based on](https://github.com/OM4/woocommerce-develop/tree/fix/37806).

Please push your already committed `includes` -> `src` changes to the branch so they are part of this PR.

I'm also more than happy for you to work on and commit/push the UI changes using the `delete_user_form` hook as well. Thank you for offering to handle that.

Konamiman commented:
@om4james Thanks, indeed now I've been able to push my commit, feel free to take a look if you want. And sure, I'll work on the UI changes.

Konamiman committed:
Add an informative message about webhooks in the "delete users" page

Konamiman committed:
Fine phpcs, yes, I'll add that comma.

Konamiman commented:
@om4james UI changes done. I've assigned a second reviewer from my team for reviewing my own changes.

thejamescollins commented on plugins/woocommerce/src/Internal/Utilities/WebhookUtil.php:
```diff
@@ -0,0 +1,136 @@
+
+		$text .= sprintf(
+			/* translators: 1 = url of the WooCommerce webhooks settings page */
+			__( 'After that they can be reassigned to the logged in user by going to the <a href="%1$s">webhooks settings page</a> and re-saving them.', 'woocommerce' ),
```
```suggestion
			__( 'After that they can be reassigned to the logged-in user by going to the <a href="%1$s">WooCommerce webhooks settings page</a> and re-saving them.', 'woocommerce' ),
```

- Explain to the user that these are WooCommerce webhooks, as the user doing the deletion may not be working in a WooCommerce context.
- Use correct `logged-in` [phrase](https://english.stackexchange.com/a/11002).

thejamescollins commented on plugins/woocommerce/src/Internal/Utilities/WebhookUtil.php:
```diff
@@ -0,0 +1,136 @@
+		// phpcs:enable WooCommerce.Commenting.CommentHooks, WordPress.DB.PreparedSQL.NotPrepared
+
+		if ( $users_have_content ) {
+			$text = __( 'If the "Delete all content" option is selected, the affected webhooks will <b>not</b> be deleted and will be attributed to user id 0.<br/>', 'woocommerce' );
```
```suggestion
			$text = __( 'If the "Delete all content" option is selected, the affected WooCommerce webhooks will <b>not</b> be deleted and will be attributed to user id 0.<br/>', 'woocommerce' );
```

- Explains to the user that these are WooCommerce webhooks, as the user doing the deletion may not be working in a WooCommerce context.

thejamescollins commented on plugins/woocommerce/src/Internal/Utilities/WebhookUtil.php:
```diff
@@ -0,0 +1,136 @@
+		if ( $users_have_content ) {
+			$text = __( 'If the "Delete all content" option is selected, the affected webhooks will <b>not</b> be deleted and will be attributed to user id 0.<br/>', 'woocommerce' );
+		} else {
+			$text = __( 'The affected webhooks will <b>not</b> be deleted and will be attributed to user id 0.<br/>', 'woocommerce' );
```
```suggestion
			$text = __( 'The affected WooCommerce webhooks will <b>not</b> be deleted and will be attributed to user id 0.<br/>', 'woocommerce' );
```

- Explains to the user that these are WooCommerce webhooks, as the user doing the deletion may not be working in a WooCommerce context.

thejamescollins commented on plugins/woocommerce/src/Internal/Utilities/WebhookUtil.php:
```diff
@@ -0,0 +1,136 @@
+			$webhooks_settings_url
+		);
+
+		echo '<p>' . esc_html( $text ) . '</p>';
```
```suggestion
		echo '<p>' . wp_kses_post( $text ) . '</p>';
```

Ensure bold and the link are output to the page, and not escaped.

Fixes this bug:

<img width="1303" alt="Screenshot 2023-05-03 at 9 01 12 am" src="https://user-images.githubusercontent.com/108233/235815519-574d14df-38d3-4379-bbfa-694010e3f8f0.png">

thejamescollins commented on plugins/woocommerce/src/Internal/Utilities/WebhookUtil.php:
```diff
@@ -0,0 +1,136 @@
+			$text = sprintf(
+				/* translators: 1 = user id, 2 = user login, 3 = webhooks count */
+				_nx(
+					'User #%1$s %2$s has created %3$d webhook.',
+					'User #%1$s %2$s has created %3$d webhooks.',
```
```suggestion
					'User #%1$s %2$s has created %3$d WooCommerce webhook.',
					'User #%1$s %2$s has created %3$d WooCommerce webhooks.',
```

- Explains to the user that these are WooCommerce webhooks, as the user doing the deletion may not be working in a WooCommerce context.

thejamescollins reviewed (commented):
Thank you @Konamiman for improving on my original idea and implementation, it looks great I think.

I have suggested one bug fix, and then also a few minor wording changes for the Delete User UI.

Thanks again.

Konamiman committed:
Improve wording, fix bad escaping of link

Co-authored-by: James Collins <james@om4.com.au>

Konamiman commented:
@om4james Thanks for the fixes, I could swear the link rendered correctly on my local setup but I must have missed something 🤷🏻‍♂️ I have committed your change suggestions.

peterfabian commented on plugins/woocommerce/src/Internal/Utilities/WebhookUtil.php:
```diff
@@ -0,0 +1,136 @@
+ */
+class WebhookUtil {
+
+	use AccessiblePrivateMethods;
```
Side note not affecting the acceptance: Do you think there should be some convention where we briefly document why this is being used and add a reference to our docs or sth like that? I'm just trying to make sure ppl understand the implications etc.

peterfabian reviewed (approved):
The reassignment works fine as described in the instructions, so 👍 from my side. Nice touch with using the WP default UI.

The only note that I have is that it's not apparent that a webhook is assigned to a(ny) user. There's nowhere in the UI that suggests this is the case, so all the user switches are done outside of the UI. Maybe adding username/userid to the webhook UI (either to the list of details) would make sense (not necessarily as part of this PR). I mean, if we're asking users to pick an API version to use for webhooks, we can let them also pick the user whose privileges should be used for the API, no? We already have the UI components in the create REST API key screen.

om4james commented:
Thank you for your help here @Konamiman and @peterfabian 💯
